product/optee-stm32mp1/module/stm32_pmic_regu/src/mod_stm32_pmic_regu.c

```diff
@@ -0,0 +1,334 @@
+         fwk_id_get_element_idx(dev_id), ctx->regu_id, *level_uv);
+
+    return FWK_SUCCESS;
+}
+
+static int pmic_regu_set_level(fwk_id_t dev_id, int level_uv)
+{
+    struct stm32_pmic_regu_dev_ctx *ctx = NULL;
+
+    ctx = module_ctx.dev_ctx_table + fwk_id_get_element_idx(dev_id);
+
+    if (!fwk_module_is_valid_element_id(dev_id)) {
+        return FWK_E_PARAM;
+    }
+
+    if (ctx->read_only) {
+        return FWK_E_ACCESS;
+    }
+
+    FWK_LOG_TRACE(MOD_NAME "SCMI voltd %u: set level PMIC %s to %d",
+         fwk_id_get_element_idx(dev_id), ctx->regu_id, level_uv);
+
+    if (set_regu_voltage(ctx->regu_id, level_uv)) {
+        return FWK_E_DEVICE;
+    }
+
+    return FWK_SUCCESS;
+}
+
+static void find_bound_uv(const uint16_t *levels, size_t count,
+                          int32_t *min, int32_t *max)
+{
+    size_t n = 0;
+
+    *min = INT32_MAX;
+    *max = INT32_MIN;
+
+    for (n = 0; n < count; n++) {
+        if (*min > levels[n]) {
+            *min = levels[n];
+        }
+        if (*max < levels[n]) {
+            *max = levels[n];
+        }
+    }
+
+    /* Convert from mV to uV */
+    *min *= 1000;
+    *max *= 1000;
+}
+
+static int pmic_regu_get_info(fwk_id_t dev_id, struct mod_voltd_info *info)
+{
+    struct stm32_pmic_regu_dev_ctx *ctx = NULL;
+    const uint16_t *levels = NULL;
+    size_t full_count = 0;
+
+    ctx = module_ctx.dev_ctx_table + fwk_id_get_element_idx(dev_id);
+
+    if (!fwk_module_is_valid_element_id(dev_id)) {
+        return FWK_E_PARAM;
+    }
+
+    stpmic1_regulator_levels_mv(ctx->regu_id, &levels, &full_count);
+
+    memset(info, 0, sizeof(*info));
+    info->name = ctx->regu_id;
+    info->level_range.level_type = MOD_VOLTD_VOLTAGE_LEVEL_DISCRETE;
+    info->level_range.level_count = full_count;
+    find_bound_uv(levels, full_count,
+                  &info->level_range.min_uv, &info->level_range.max_uv);
+
+    FWK_LOG_TRACE(MOD_NAME "SCMI voltd %u: get_info PMIC %s, range [%d %d]",
+         fwk_id_get_element_idx(dev_id), ctx->regu_id,
+         info->level_range.min_uv, info->level_range.max_uv);
+
+    return FWK_SUCCESS;
+}
+
+static int pmic_regu_level_from_index(fwk_id_t dev_id, unsigned int index,
+                                      int32_t *level_uv)
+{
+    struct stm32_pmic_regu_dev_ctx *ctx = NULL;
+    const uint16_t *levels = NULL;
+    size_t full_count = 0;
+
+    ctx = module_ctx.dev_ctx_table + fwk_id_get_element_idx(dev_id);
+
+    if (!fwk_module_is_valid_element_id(dev_id)) {
+        return FWK_E_PARAM;
+    }
+
+    stpmic1_regulator_levels_mv(ctx->regu_id, &levels, &full_count);
+    if (index >= full_count) {
+        return FWK_E_RANGE;
+    }
+
+    *level_uv = (int32_t)levels[index] * 1000;
+
+    FWK_LOG_TRACE(MOD_NAME "SCMI voltd %u: get level PMIC %s = %d",
+         fwk_id_get_element_idx(dev_id), ctx->regu_id, *level_uv);
+
+    return FWK_SUCCESS;
+}
+
+static const struct mod_voltd_drv_api api_stm32_pmic_regu = {
```

Comment:
For ease of finding things, maybe order the functions above this in the order they appear within this struct